### PR TITLE
feat: delete unprocessed entity from refresh_state table

### PR DESCRIPTION
**Pull request by kurtaking**
## Hey, I just made a Pull Request!

re: https://github.com/backstage/backstage/issues/21858

Add the ability to delete a record from the `refresh_state` table.

image

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] Added or updated documentation
- [ ] Tests for new functionality and regression tests for bug fixes
- [x] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

#### Remaining todos

- [x] add support for permissions
- [x] determine - do we need this functionality on both the `Failed` and `Pending` tabs?
- [x] address breaking change concerns

### CODE REVIEW

**Comment by awanlin** [comment]
@kurtaking, this is awesome! I know this is Draft but just want to mention that this should probably have support for being locked down with permissions. 👍

**Comment by kurtaking** [comment]
I could use feedback on what we want to do with the `<PendingEntities />` component.

File: plugins/catalog-unprocessed-entities/src/components/FailedEntities.tsx
Comment (kurtaking) [inline plugins/catalog-unprocessed-entities/src/components/FailedEntities.tsx:116]: todo: pass in `entityRef`

**Comment by benjdlambert** [comment]
What's the use case for deleting `unprocessed_entities`? 

I'm might be misunderstanding here, but I think that having many `unprocessed_entities` is a symptom of a larger issue that the things on the queue are not being processed in a timely manner. They should eventually be moved from there eventually I believe. @freben to confirm?

**Comment by freben** [comment]
Not quite, could also be that they are in a corner case where they never successfully get processed or so, for example.

**Comment by kurtaking** [comment]
hey @benjdlambert, does the info in this [issue](https://github.com/backstage/backstage/issues/21858) provide enough context?

**Comment by kurtaking** [comment]
I'm pausing this work until I hear back.

**Comment by awanlin** [comment]
@kurtaking, please keep going, I think this is a totally valid feature to add and I know it would have helped a few people recently on Discord. 👍

**Comment by alde** [comment]
I like this, the plugin was meant as a bit of an admin tool to get some insights into why entities don't show up, but it certainly could make sense to be able to remove things too. Worst case scenario if you delete something that's still tracked it'll come back with the same issue, so not really a big problem I believe.

**Comment by alde** [comment]
> I could use feedback on what we want to do with the <PendingEntities /> component (tab).

Pending entities I don't really think should be deleted - they should either get processed or failed. Ideally they shouldn't be pending for very long.

**Comment by kurtaking** [comment]
hey @awanlin, any recommendation on resolving the failed **Preview (build) / Build PR image (pull_request)** step?

**Comment by awanlin** [comment]
Honestly not sure but I triggered it to run again and we'll see how that goes 🤔

**Comment by awanlin** [comment]
Looks like that did it, builds are green! 🚀

**Comment by benjdlambert** [comment]
@alde gonna leave this for you and @backstage/catalog-maintainers to review and ship :tada:

File: .changeset/warm-buses-do.md
Comment (benjdlambert) [inline .changeset/warm-buses-do.md:8]:
```suggestion

```

File: plugins/catalog-backend-module-unprocessed/api-report.md
Comment (benjdlambert) [inline plugins/catalog-backend-module-unprocessed/api-report.md:19]: This is a breaking change now, so it should be called out as such if we can't make it backwards compatible.

**Review by benjdlambert — COMMENTED** [review]
Just some quick bits here.

File: plugins/catalog-backend-module-unprocessed/api-report.md
Comment (kurtaking) [reply plugins/catalog-backend-module-unprocessed/api-report.md:19]: Ah, breaking because it requires the end user to add permissions. I like the idea of making it backwards compatible.

Comment (kurtaking) [reply plugins/catalog-backend-module-unprocessed/api-report.md:19]: I'll do some searching on my own - do any examples exist today where implementing permissions in a plugin is optionally provided?

Comment (awanlin) [reply plugins/catalog-backend-module-unprocessed/api-report.md:19]: Umm, DevTools ;)

Comment (alde) [reply plugins/catalog-backend-module-unprocessed/api-report.md:19]: I would say that if you can, don't break existing APIs.
Should maybe also just not allow deletions if you're not passing in permissions, and return an error that guides people to how to fix it.

Comment (kurtaking) [reply plugins/catalog-backend-module-unprocessed/api-report.md:19]: I need help adding permissions without it being a breaking change, and I don't see how permissions are optional in DevTools. 😞 I have to pass in permissions as options to the router. Is the right decision to make the entire `RouterOptions` object optional? Users will still have to pass in an empty object if I make permissions optional, correct? That's a breaking change 🤔. 

```ts
interface RouterOptions = {
  permissions?: PermissionsEvaluator
}
```

Comment (awanlin) [reply plugins/catalog-backend-module-unprocessed/api-report.md:19]: Sorry, permissions are optional as far as creating the rules for them in DevTools but the `PermissionsEvaluator` is required and I'm not seeing any good examples where it's not :(

Comment (kurtaking) [reply plugins/catalog-backend-module-unprocessed/api-report.md:19]: Our internal version of this isn't wrapped with permissions 🙃 we didn't see it as a significant enough threat since the table repopulates when there is an entity with a valid error. 

I don't see a way to ship without breaking change. Should I strip out permissions and 🚢, or make it a breaking change? 

cc: @alde

Comment (alde) [reply plugins/catalog-backend-module-unprocessed/api-report.md:19]: I think a breaking change isn't always bad, people can choose not to upgrade

**Comment by kurtaking** [comment]
> @kurtaking given a breaking change isn't a deal breaker, is there anything left to do?

@alde, I was out moving last week. I will update the changelog to be a major bump and we should be good to go 👍🏼

**Comment by benjdlambert** [comment]
@kurtaking think that the changeset should still be `minor` if it's a breaking change. The package is still `0.x` I think, so `minor` is effectively for breaking changes when the package is less than stable.

**Comment by kurtaking** [comment]
> @kurtaking think that the changeset should still be `minor` if it's a breaking change. The package is still `0.x` I think, so `minor` is effectively for breaking changes when the package is less than stable.

@benjdlambert - TIL, thank you! I'll get that updated.

File: .changeset/warm-buses-do.md
Comment (alde) [inline .changeset/warm-buses-do.md:9]: Love the clear message in the changeset

**Review by alde — APPROVED** [review]

**Comment by kurtaking** [comment]
Looking for another review from the maintainers 👋🏼 ❤️

**Comment by kurtaking** [comment]
@vinzscam, I see you are listed as codeowner under the reviewers. Can you help with the second review?

**Comment by kurtaking** [comment]
@awanlin, can you help get me traction with the required reviews?

**Comment by benjdlambert** [comment]
@kurtaking can you rebase the PR? :pray:

**Comment by kurtaking** [comment]
> @kurtaking can you rebase the PR? 🙏

@benjdlambert, doing so is causing some unexpected failures.

**Comment by benjdlambert** [comment]
@kurtaking I think I might have fixed it 🤞

**Comment by kurtaking** [comment]
> @kurtaking I think I might have fixed it 🤞

@benjdlambert, your update contained a few errors, so I rebased it again. Still running into issues 😢

**Comment by kurtaking** [comment]
thank you, @benjdlambert!! 🙌🏼

File: .changeset/warm-buses-do.md
Comment (kurtaking) [inline .changeset/warm-buses-do.md:1]: This is much better, too. Definitely taking away some learnings from the PR. 🙏🏼

**Review by benjdlambert — APPROVED** [review]

File: plugins/catalog-backend-module-unprocessed/src/UnprocessedEntitiesModule.ts
Comment (benjdlambert) [inline plugins/catalog-backend-module-unprocessed/src/UnprocessedEntitiesModule.ts:124]: @camilaibs does this need converting to use the new auth APIs? Or do permissions calls just forward the token blindly?

Comment (freben) [reply plugins/catalog-backend-module-unprocessed/src/UnprocessedEntitiesModule.ts:124]: `.authorize` [now accepts `credentials` directly](https://github.com/backstage/backstage/blob/bf2c435983b18cc6674c226b82d2694557e53252/plugins/azure-sites-backend/src/service/router.ts#L115), instead of a token

**Review by vinzscam — APPROVED** [review]

**Comment by benjdlambert** [comment]
@kurtaking I just want to make the changes to support the new auth system with this before shipping, apologies for the delay, will try and get to this in the next day or so :pray:

**Review by benjdlambert — DISMISSED** [review]
Just gonna block this so it doesn't get accidentally merged, but will try and get this in the release for tomorrow.

**Comment by benjdlambert** [comment]
@kurtaking I think this should be good to go now. Sorry for the delay - I've also just updated the changeset to reflect a little update in the APIs. Thought it was better to wrap this up in a `create` static method similar to the `.fromConfig` statics that we have so that we can evolve the API a little easier rather than using a constructor for this.

**Review by benjdlambert — APPROVED** [review]

**Comment by kurtaking** [comment]
hey @benjdlambert, thank you for helping get this shipped. This was a great experience for me, and the code review taught me a lot. I appreciate it!

**Comment by webark** [comment]
@kurtaking and @benjdlambert I'm seeing an issue where the

https://github.com/backstage/backstage/blob/9dbf9f6cf4ae28352e25a069467d65316bf6762e/plugins/catalog-backend-module-unprocessed/src/UnprocessedEntitiesModule.ts#L165

is turning "owner" into a string of "undefined", so it never returns anything.

Am I missing something with my setup? Or is this a bug?

This is due to `String({}.foo) === 'undefined'`.

**Comment by webark** [comment]
yea, if I add `?owner=` to the url like `/api/catalog/entities/unprocessed/failed?owner=` it returns me back the full list of what I want.

**Comment by kurtaking** [comment]
> yea, if I add `?owner=` to the url like `/api/catalog/entities/unprocessed/failed?owner=` it returns me back the full list of what I want.

Hey @webark, thanks for bringing this up. I can look at this over the weekend, but I am slammed this week. Looks like we need to do additional testing around [this change](https://github.com/backstage/backstage/commit/0c234a73ac4a59699b492ce82050f4c95a4e73e6).

**Comment by awanlin** [comment]
@kurtaking CC @webark - I think this PR is working towards a fix - title makes it hard to realize thats the case though: https://github.com/backstage/backstage/pull/24519

**Pull request merged**
